pkgs/servers/klipper/klipper-firmware.nix

```diff
@@ -0,0 +1,62 @@
+{ stdenv
+, lib
+, pkg-config
+, pkgsCross
+, bintools-unwrapped
+, libffi
+, libusb1
+, wxGTK
+, python2
+, python3
+, gcc-arm-embedded
+, klipper
+, avrdude
+, stm32flash
+, mcu ? "mcu"
+, firmwareConfig ? ./simulator.cfg
+, flashDevice ? null
+}: stdenv.mkDerivation rec {
+  name = "klipper-firmware-${mcu}-${version}";
+  version = klipper.version;
+  src = klipper.src;
+
+  nativeBuildInputs = [
+    python2
+    python3
+    pkgsCross.avr.stdenv.cc
+    gcc-arm-embedded
+    bintools-unwrapped
+    libffi
+    libusb1
+    avrdude
+    stm32flash
+    pkg-config
+    wxGTK # Required for bossac
+  ];
+
+  preBuild = "cp ${firmwareConfig} ./.config";
+
+  postPatch = ''
+    patchShebangs .
+  '';
+
+  makeFlags = [
+    "V=1"
+    "KCONFIG_CONFIG=${firmwareConfig}"
+  ];
+
+  installPhase = ''
+    mkdir -p $out
+    cp ./.config $out/config
+    cp -r out/* $out
+  '' + lib.optionalString (flashDevice != null) "make FLASH_DEVICE=${toString flashDevice} OUT=$out/ KCONFIG_CONFIG=$out/config flash";
```

Comment:
I don't think we should attempt to flash during the package build process here.

Reply:
Yes I had the same doubts, which is why I implemented this configurably:
- By default `flashDevice` is `null`. Flashing does not happen here if you just install the firmware. You have to actively override the `flashDevice`. You have to override it anyways if you don't want to use the klipper service.
- The main intended usage is via the klipper service: Defining `services.klipper.firmware.mcu.firmwareConfig` will build the firmware but not flash it. But if you want to flash the firmware on rebuilds, you can enable it via `services.klipper.firmware.mcu.enableFlashing = true`. 

With this, I think everyone will be happy. I personally consider the automatic flashing, as my microcontroller is also part of the build process.

Reply:
Flashing a device inside of a Nix build is a hard no. Absolutely not.

The right way of doing this is:
1. Write a derivation for the firmware
2. Add a trigger to the systemd unit that calls the flashing command.

Reply:
I absolutely understand you concerns. By default the `flashing.enable` option is off and the description of the service has a big `WARNING`. In all cases without intention the nix-build will not flash anything. Hopefully we can leave this option.

One of my main motivation for having it in Nix build is because of the flashing process of Klipper. Various firmwares require additional tools like `bossac` or `hidflash`. Klipper actually builds them when you flash 😅 .
It is really difficult in Nix to move this flashing outside of the Nix build. That is also the reason the `klipper-flash` script does not support some boards. Additionally, klipper actually copy pasted those tools into its own repository and uses them there with hardcoded paths 😅 
So to flash outside of buildprocess for those various boards, you have to heavily modify the Klipper source which is in the Nix store and readonly.

Regarding your suggestion: 
1. I think this is done. 
2. I'm unsure what you mean here. Should I create an additional `oneshot` systemd unit for flashing? That would just start and then would have the same effect as "uncontrollably" flashing on Nix build. Or how would I set it to `disabled` state by default so I can trigger it manually?

Reply:
Nix builds should never cause side-effects: They only take build inputs and put files in designated output paths in `/nix/store`. Only the user can cause side-effects, such as by running the activation script in a NixOS system profile thereby making the config actually take effect. Until the configuration is activated, it's entirely "static" and doesn't affect the operation of the running system. What you have here doesn't work in a normal sandboxed setup, and builds don't always happen on the host connected to the printer (e.g., with deployment tools like NixOps and Colmena).

If it's difficult to separate the flash routine for those tricky arches, then we shouldn't do it at all at the moment. Let's get `klipper-flash` in first, and maybe send patches upstream to make things better. A big warning doesn't make it acceptable - It's inconsistent with how Nix works.

> Should I create an additional oneshot systemd unit for flashing? That would just start and then would have the same effect as "uncontrollably" flashing on Nix build.

Services don't get started unless you add `wantedBy = [ "multi-user.target" ];` or otherwise insert dependencies from other units.

Reply:
This sounds reasonable!
I will remove the `flashing.enable` setting, but is a second service or some kind of service involvement really necessary? I mean just calling `klipper-flash-mcu` does the job currently as well.

Reply:
I changed the `flashing.enable` option to a simple `enable` option. Only when enabled the firmwares and its flashing scripts are added and built.